File: ios/NorthAx/Services/AuthService.swift
import Foundation
import Observation

/// Owns the authenticated session. Email/password credentials are exchanged with
/// the backend (`POST /auth/login` or `POST /auth/register`) for an access +
/// refresh JWT pair stored in the Keychain (§3). The session is restored on
/// launch from those tokens, and dropped if the backend revokes them.
@MainActor
@Observable
class AuthService {
    private(set) var currentUser: AuthUser? = nil
    var authError: AuthSignInError? = nil
    private(set) var isAuthenticating: Bool = false

    private let api = APIClient.shared
    private let tokens = TokenStore.shared
    private let cachedUserKey = "northax.cachedUser"
    private var expiryObserver: NSObjectProtocol?

    init() {
        observeSessionExpiry()
        restoreSession()
    }

    var isAuthenticated: Bool { currentUser != nil }

    // MARK: - Email / Password

    /// Sign in with an existing account (§3.1).
    func signIn(email: String, password: String) {
        let email = email.trimmingCharacters(in: .whitespacesAndNewlines)
        guard Self.isValidEmail(email) else {
            authError = .invalidInput("Enter a valid email address.")
            return
        }
        guard !password.isEmpty else {
            authError = .invalidInput("Enter your password.")
            return
        }
        authError = nil
        Task { await exchange(path: "auth/login", body: EmailSignInRequest(email: email, password: password)) }
    }

    /// Create a new account, then start an authenticated session (§3.1).
    func register(name: String, email: String, password: String) {
        let name = name.trimmingCharacters(in: .whitespacesAndNewlines)
        let email = email.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else {
            authError = .invalidInput("Enter your name.")
            return
        }
        guard Self.isValidEmail(email) else {
            authError = .invalidInput("Enter a valid email address.")
            return
        }
        guard password.count >= 8 else {
            authError = .invalidInput("Password must be at least 8 characters.")
            return
        }
        authError = nil
        Task {
            await exchange(path: "auth/register", body: EmailSignUpRequest(name: name, email: email, password: password))
        }
    }

    /// Exchange credentials for app tokens and open the session.
    private func exchange(path: String, body: Encodable) async {
        isAuthenticating = true
        defer { isAuthenticating = false }
        do {
            let resp: AuthResponse = try await api.post(path, body: body, authenticated: false)
            tokens.save(accessToken: resp.accessToken, refreshToken: resp.refreshToken)
            setUser(AuthUser(id: resp.user.id, name: resp.user.name, email: resp.user.email))
        } catch let error as APIError {
            authError = .server(error.userMessage)
        } catch {
            authError = .failed
        }
    }

    private static func isValidEmail(_ email: String) -> Bool {
        // Pragmatic check: one @, non-empty local part, dotted domain.
        email.range(of: #"^[^@\s]+@[^@\s]+\.[^@\s]+$"#, options: .regularExpression) != nil
    }

    // MARK: - Sign Out

    func signOut() {
        // Best-effort server revoke; local clear is authoritative.
        Task { _ = try? await api.send("DELETE", "auth/session") }
        tokens.clear()
        UserDefaults.standard.removeObject(forKey: cachedUserKey)
        currentUser = nil
    }

    // MARK: - Session restore

    private func restoreSession() {
        guard tokens.hasSession else { return }
        // Optimistic restore from cache, then validate against the backend.
        if let data = UserDefaults.standard.data(forKey: cachedUserKey),
           let user = try? JSONDecoder().decode(AuthUser.self, from: data) {
            currentUser = user
        }
        Task {
            do {
                let profile: UserProfileDTO = try await api.get("user/profile")
                setUser(AuthUser(id: profile.id, name: profile.name, email: profile.email))
            } catch let error as APIError where error.isUnauthorized {
                signOut()
            } catch {
                // Network error: keep optimistic session; will retry next launch.
            }
        }
    }

    private func setUser(_ user: AuthUser) {
        currentUser = user
        if let data = try? JSONEncoder().encode(user) {
            UserDefaults.standard.set(data, forKey: cachedUserKey)
        }
    }

    private func observeSessionExpiry() {
        expiryObserver = NotificationCenter.default.addObserver(
            forName: .northaxSessionExpired, object: nil, queue: .main
        ) { [weak self] _ in
            MainActor.assumeIsolated {
                guard let self else { return }
                self.currentUser = nil
                UserDefaults.standard.removeObject(forKey: self.cachedUserKey)
            }
        }
    }

    // MARK: - Debug bypass

#if DEBUG
    /// Local-only session for offline UI work. Has no backend tokens, so live
    /// API calls will not succeed — the app should fall back to client engines.
    func signInAsDebugUser(name: String = "Rafael") {
        setUser(AuthUser(id: "debug-\(UUID().uuidString)", name: name, email: nil))
        authError = nil
    }
#endif
}

// MARK: - Error type

enum AuthSignInError: LocalizedError {
    case invalidInput(String)
    case failed
    case unknown
    case server(String)

    var errorDescription: String? {
        switch self {
        case .invalidInput(let message):
            return message
        case .failed:
            return "Sign in failed. Please try again."
        case .unknown:
            return "Something went wrong. Please try again."
        case .server(let message):
            return message
        }
    }
}

File: android/app/src/main/java/app/northax/store/AuthService.kt
package app.northax.store

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.northax.data.AppContainer
import app.northax.data.remote.ApiError
import app.northax.data.remote.dto.AuthResponse
import app.northax.data.remote.dto.EmailSignInRequest
import app.northax.data.remote.dto.EmailSignUpRequest
import app.northax.data.remote.dto.UserProfileDto
import app.northax.domain.model.AuthUser
import kotlinx.coroutines.launch

/**
 * Owns the authenticated session. Email/password credentials are exchanged
 * with the backend for an access + refresh JWT pair stored encrypted. The
 * session is restored on launch from those tokens, and dropped if the backend
 * revokes them — a 1:1 port of the iOS AuthService.
 */
class AuthService(private val container: AppContainer) : ViewModel() {

    var currentUser by mutableStateOf<AuthUser?>(null)
        private set
    var authError by mutableStateOf<String?>(null)
    var isAuthenticating by mutableStateOf(false)
        private set

    private val client = container.apiClient
    private val tokens = container.tokens
    private val prefs = container.prefs

    val isAuthenticated: Boolean get() = currentUser != null

    init {
        observeSessionExpiry()
        restoreSession()
    }

    // MARK: - Email / Password

    /** Sign in with an existing account. */
    fun signIn(email: String, password: String) {
        val trimmed = email.trim()
        if (!isValidEmail(trimmed)) {
            authError = "Enter a valid email address."
            return
        }
        if (password.isEmpty()) {
            authError = "Enter your password."
            return
        }
        authError = null
        viewModelScope.launch { exchange("auth/login") { post(it, EmailSignInRequest(trimmed, password)) } }
    }

    /** Create a new account, then start an authenticated session. */
    fun register(name: String, email: String, password: String) {
        val trimmedName = name.trim()
        val trimmedEmail = email.trim()
        if (trimmedName.isEmpty()) {
            authError = "Enter your name."
            return
        }
        if (!isValidEmail(trimmedEmail)) {
            authError = "Enter a valid email address."
            return
        }
        if (password.length < 8) {
            authError = "Password must be at least 8 characters."
            return
        }
        authError = null
        viewModelScope.launch {
            exchange("auth/register") { post(it, EmailSignUpRequest(trimmedName, trimmedEmail, password)) }
        }
    }

    private suspend fun post(path: String, body: EmailSignInRequest): AuthResponse =
        client.post(path, body, authenticated = false)

    private suspend fun post(path: String, body: EmailSignUpRequest): AuthResponse =
        client.post(path, body, authenticated = false)

    /** Exchange credentials for app tokens and open the session. */
    private suspend fun exchange(path: String, request: suspend (String) -> AuthResponse) {
        isAuthenticating = true
        try {
            val resp = request(path)
            tokens.save(resp.accessToken, resp.refreshToken)
            setUser(AuthUser(resp.user.id, resp.user.name, resp.user.email))
        } catch (e: ApiError) {
            authError = e.userMessage
        } catch (_: Exception) {
            authError = "Sign in failed. Please try again."
        } finally {
            isAuthenticating = false
        }
    }

    // MARK: - Sign Out

    fun signOut() {
        // Best-effort server revoke; local clear is authoritative.
        viewModelScope.launch {
            try {
                client.send("DELETE", "auth/session")
            } catch (_: Exception) {
            }
        }
        tokens.clear()
        prefs.cachedUser = null
        currentUser = null
    }

    // MARK: - Session restore

    private fun restoreSession() {
        if (!tokens.hasSession) return
        // Optimistic restore from cache, then validate against the backend.
        prefs.cachedUser?.let { currentUser = it }
        viewModelScope.launch {
            try {
                val profile: UserProfileDto = client.get("user/profile")
                setUser(AuthUser(profile.id, profile.name, profile.email))
            } catch (e: ApiError) {
                if (e.isUnauthorized) signOut()
                // Other errors: keep optimistic session; will retry next launch.
            } catch (_: Exception) {
            }
        }
    }

    private fun setUser(user: AuthUser) {
        currentUser = user
        prefs.cachedUser = user
    }

    // MARK: - Debug bypass

    /** Local-only session for offline UI work (debug builds only). Has no
     *  backend tokens, so live API calls will not succeed — the app falls back
     *  to the client engines. */
    fun signInAsDebugUser(name: String = "Rafael") {
        setUser(AuthUser("debug-${java.util.UUID.randomUUID()}", name, null))
        authError = null
    }

    private fun observeSessionExpiry() {
        viewModelScope.launch {
            client.sessionExpired.collect {
                currentUser = null
                prefs.cachedUser = null
            }
        }
    }

    private companion object {
        // Pragmatic check: one @, non-empty local part, dotted domain.
        val emailRegex = Regex("^[^@\\s]+@[^@\\s]+\\.[^@\\s]+$")
        fun isValidEmail(email: String): Boolean = emailRegex.matches(email)
    }
}
